### DISCUSSION TITLE
Impossible to find recovery files

### DISCUSSION THREAD

**Discussion by Soothfast** [xournalpp/xournalpp#5325] [Q&A]
Here's another issue with Xournal++ on Ubuntu.  I can think of no other program or app where the "recovery" system itself poses the greatest threat of massive and permanent loss of data, both on Windows and now (tonight) on Ubuntu.  If I don't FORCE the "recovery file" to save on, say, the desktop, it's gone.  

10 minutes into a lecture, and Xournal++ crashed again and again whenever I tried to cut and paste a bit of typed text from one page to another.  Giving up on trying the cut and paste, I moved on, naming the "recovery file" something and continuing with writing notes for 90 minutes.  I hit save manually a lot during this, and then I closed out the file.  The file is on the desktop.  Or is it!?  Nope:  the "recovery file" must be hidden somewhere, because the file on the desktop had nothing of that 90 minutes of writing.  

I searched everywhere.  In all the folders instructions say to look, there's nothing.  I tried that cut and paste job again, bringing about another crash, and this time I looked for the lost file on the window that pops up for making a recovery file.  The missing file was there, in the very folder that's actually empty when I reach it by other means, but in this pop-up window I cannot touch it.  So the file maybe doesn't really exists after all, and I'm looking at a "ghost" file?

I have no understanding of this recovery file thing, as I've never encountered another program that handles crash recovery in such a peculiar way.  It seems dangerous, unreliable, complicated, and unintuitive.  Why couch a recovery file in a whole new and obscure place?  With the last crash, I forced the recovery file to be saved on the desktop.  I then hit the save button multiple times.  I then shut down Xournal.  I then opened the recovery file and argued with a new prompt to send a bug report and/or make a recovery file out of the recovery file.  Closed Xournal again.  Reopened it and swatted away another prompt for making a bug report (sending the report apparently requires special software).  Finally verified no data was lost.  Closed and reopened the file yet again to ensure I was not going to be harassed about a bug report until the end of time.  Then deleted the original file.  Then gave the "recovery file" the name of the original file I just deleted.  And it was done.

So do I have a "ghost" file, or is the file hidden somewhere?  A search does not turn up the missing file.  Is Ubuntu's search tool by default only limited to some subsection of the hard drive, like the documents folder?  Is there a way to FORCE Xournal to save recovery files on the desktop by default?

Where I went to find a recovery file is here:
home/[user name]/snap/xournalpp/
There's folders 86, common, and current.  All empty.

I do not have a cache folder under home/[user name].  Never did.  

root/snap/xournalpp is another path leading to empty folders 86, common, and current.  Same place?  Not sure.

If this comes to no resolution, then I must be sure to save the recovery file **anywhere** but where Xournal wants to stow it, or else bye-bye data.

**Comment by rolandlo** [comment]
I'm sorry for the bad experience you got while using Xournal++. Let's try to understand what happened and what can be done in future.

Xournal++ (snap version) has crashed and on restart opened the `emergencysave.xopp` inside the config folder. For the snap version this is `~/snap/xournalpp/current/.config/xournalpp`, which is somehow undocumented. Here `current` is a symlink to the actual folder (whose name is something like `x<revision number>`, e.g. `x17`). 
You may have missed that folders starting with a dot like `.config` are hidden by default in the file explorer (`nautilus`). You can hit `CTRL+h` to make those hidden folders and files visible.

When you renamed the emergency file, you stayed inside the same folder, so the file got stored in `~/snap/xournalpp/current/.config/xournalpp`, which you couldn't find anymore.

The resolution is simple: Whenever you recover from an emergency save xopp-file, choose menu `File > Save As`, switch folders to where you really want the file to be saved, choose a file name and only then hit enter.

From our side we should document [where to find files](https://xournalpp.github.io/guide/file-locations/) when running the snap package or the flatpak. Also I think the `emergencysave.xopp` should be stored in the cache folder, not the config folder.

About the bugreport dialog: There were some issues with crashlog files not opening up correctly, but it has been fixed in version 1.2.1. Make sure you run the latest version of the snap package. If you switch to the candidates or edge channel, you get version 1.2.2. The dialog should show the options to
- send the bugreport (which should open the web-browser and display the bugreport template while also opening the logfile in a text editor)
- open the logfile (in a text editor)
- open the logfile directory (in the file manager)
- delete the logfile
- cancel

Note that the (crash/error)logfile is not the same thing as the `emergencysave.xopp` file. The latter is the Xournal++ document, the former just logs how the crash came about (it's contains a backtrace, listing the commands that were executed right before Xournal++ crashed). The logfiles help us to fix bugs in Xournal++, while the `emergencysave.xopp` helps you recover your notes.

The `emergencysave.xopp` file is always stored at the same (slightly obscure) location, because the only thing for Xournal++ to do on startup is to check if that file exists at the given location.

There are also `autosave` files (with a name like `.foo.autosave.xopp`) that get stored in the same folder as your original `xopp`-file. Depending on the autosave time interval, they may be slightly incomplete though. They also start with a dot, so they are hidden by default as well.

**Comment by Soothfast** [comment]
Thanks for the reply.

I lost half of a trigonometry lecture a few days ago to this (to me) very bizarre "recovery" system, despite my practice of manually doing Ctrl+S every few minutes.  I've never lost so much work as I do using this program.  Perhaps I do not really have the "snap" version.  Is there a way to definitively divine what exact version I have?  I see the xournalpp folder inside the snap folder, so surely that confirms I have the snap package (v1.2.1 currently).

Last crash, I chose the option to get the recovery file.  Xournal could not find it.  So I really don't believe there exist any "hidden" files.

I have things set to autosave every 2 minutes.  I keep my xopp files on the desktop.  Does this not mean the program is doing Ctrl+S just as I do, for the document on the _desktop??_  If not, then what's the use?  And if so, how can I lose 45 minutes of work?

By the way, here's what causes about two-thirds of my crashes:  trying to copy/paste something more than 8 or 10 pages back.  It just flips out.  The Windows version does not do this.

Is there a way to COMPLETELY disable the auto-recovery nonsense so I can just do saves manually every few minutes?  What if I just uncheck autosave in the preferences?  Would that do it?  I want all bridges to these mystical recovery folders burned to the ground.  Having recently learned what a mess the Linux community is in (over 300 versions of Linux and counting), I must assume I have some "flavor" of Ubuntu that is just not quite totally compatible with the snap version of Xournal.

**Reply by rolandlo** [reply]
@Soothfast I'm sorry for the bad experience you got. Maybe you can still save your 45 minutes of lecture notes though.

It would definitely help to know which version you have. From your description it sounds like you are on some Ubuntu flavour using the snap version 1.2.1. You get Xournal++'s version from menu `Help > About`.
You also get it from running `xournalpp --version` or (if you have the snap package) `snap run xournalpp --version` in the terminal.

The crashes you encounter are fixed in version >= 1.2.2. If you want to continue to use the snap version, just switch to the edge or candidate channel. It's not our packaging and we are currently waiting for @kenvandine to merge the version bump to 1.2.3. 

Otherwise you could also download an [official release](https://github.com/xournalpp/xournalpp/releases), either version 1.2.3 or the nightly build. Just make sure you get the version that fits to your Ubuntu distribution.

No, autosave does not do the same as Ctrl+S at all. It creates a backup of the document you work on and doesn't overwrite the current file.

Assuming version 1.2.1 you might find the backup either in the same desktop folder that you saved your file or in the "mystical" `.cache` folder (`~/.cache/xournalpp/autosaves`). You also find this info on the [website](https://xournalpp.github.io/guide/file-locations/#where-to-find-xournal-files).

You can of course disable autosaving in menu `Edit > preferences`, but I would not recommend doing so.

**Reply by kenvandine** [reply]
1.2.2 is now in the stable channel and 1.2.3 will be in the edge channel soon.  Once that gets sufficient testing I'll promote it to stable as well.

**Reply by kenvandine** [reply]
1.2.3 has now been published to the stable channel of the snap store.

**Reply by Soothfast** [reply]
Thanks again for the replies.  As I said in November I have no .cache folder.  I did not delete it:  it was never in existence.

I have no idea what the edge or candidate channel is.  I'm a "casual user" and come by here quite infrequently.

When I run `snap run xournalpp --version` I get this:

Not loading module "atk-bridge": The functionality is provided by GTK natively. Please try to not load it.
xournalpp 1.2.2 (09ae3333)
└──libgtk: 3.24.38

Something very odd has happened.  Yesterday I had version 1.2.1, and today it now says 1.2.2.  I have no explanation for this, because I did not set out to update the app.  In my experience updating things in Linux is a real chore and anything but automatic, but my loathing for Apple and Microsoft keep me motivated to stick with it. :)

Was v1.2.2 only just put on the "stable channel" in the past 24 hours?  I had the impression this version has been around for some time now.

Anyway it's done, and I have v1.2.2.  But I still don't have a .cache folder, and have no inkling what that portends.  I'll see how things go.

Regarding my loss of 45 minutes of work, my guess is that I made a fatal error when, after the app crashed last week, I opted to replace the crashed file with a file of the same name in the same folder.  At the time, all but the last 30 seconds of my work was still in existence.  The loss occurred later somehow, when the file was closed and reopened.  And later still, about 5% of the work returned, like some kind of partial recovery was trying to happen while I slept.  Really weird.  But if the app cannot find the recovery file, what hope have I of finding it?

Do I need to create a .cache folder manually?

**Comment by rolandlo** [comment]
> Thanks again for the replies. As I said in November I have no .cache folder. I did not delete it: it was never in existence.
> 
I'm wondering how you concluded that you don't have a .cache folder. Did you simply open the file manager (nautilus on Ubuntu) and see there? Note that folders starting with a `.` are hidden by default. You would have to press `CTRL+h` in nautilus to unhide them. Alternatively you could type 
```term
ls ~/.cache
``` 
in the terminal to list this directory (the `ls` command is for listing files and directories). For the snap package the cache folder is  `~/snap/xournalpp/common/.cache/xournalpp/` though, so you would have to look there. I should have pointed that out earlier, sorry about that.

> I have no idea what the edge or candidate channel is. I'm a "casual user" and come by here quite infrequently.
> 
If you use the software center on Ubuntu to install software (which I guess you did) you can see the different channels in the dropdown you can see on the screenshot.
![grafik](https://github.com/xournalpp/xournalpp/assets/40485433/daaf3c56-57ae-4e3d-a36b-985c3cd5799d)

If you install from command line you can also switch to a different channel using

```term
snap switch xournalpp --edge
snap refresh
```

As you see currently all three channels are on version 1.2.3, so it currently doesn't make a difference.
> 
> Something very odd has happened. Yesterday I had version 1.2.1, and today it now says 1.2.2. I have no explanation for this, because I did not set out to update the app. In my experience updating things in Linux is a real chore and anything but automatic, but my loathing for Apple and Microsoft keep me motivated to stick with it. :)
>
Snap package are updated automatically in contrast to other packages on Linux.
 
> Was v1.2.2 only just put on the "stable channel" in the past 24 hours? I had the impression this version has been around for some time now.
> 
Yes, it was just very recently.

> 
> Do I need to create a .cache folder manually?

No you don't. See above.

**Reply by ghoshsrikrishna** [reply]
Thank you very much @rolandlo . You saved me. The xournal app is probably the best app I have been using in the last month. I had similar issues which I could resolve following your comments.

**Comment by Soothfast** [comment]
All right, well, in my Home folder there is a .cache folder -- but that's not the right .cache folder, is it?  I had the impression we're talking about a .cache file in Home --> snap --> xournalpp, which I don't seem to have (or can't see).  Is it possible to see .cache folders in some folders but not in others?

With apologies, I can't seem to find where Nautilus is hidden.  Under Home --> .config there is a folder called nautilus, which contains nothing aside from a file called search-metadata.  I've scoured the Internet, but the documentation out there is incredibly user-hostile and/or out-of-date and/or flat wrong ("open nautilus with Ctrl + Alt + Home" -- nope!)  Can it only be accessed through the terminal??

**Comment by Soothfast** [comment]
I've now been updated to v1.2.3 at some point, automatically.  And now the Latex tool is broken.  Great.

**Reply by Soothfast** [reply]
I will say, though, that with v1.2.3 I can find the autosave folder, and it has stuff in it!

Alas, that aforementioned missing 45 minutes was not among the autosaves.  I think I know why, so next time there's a crash I hope to be smarter about recovering.

**Comment by Soothfast** [comment]
Good news:  I found my missing 45 minutes of work!  It was not in the autosaves folder.  A couple autosaves (yes, just 2) ended up in snap/xournalpp/94/.config/xournalpp.  I have no explanation for this.  Perhaps xournalpp is not completely compatible with the version of Ubuntu that comes preinstalled on a ThinkPad X1.

**Comment by rolandlo** [comment]
In that case you must have used a version prior to version 1.2.0 (released on July 30, 2023), maybe 1.1.3. See how important it is to tell what version you used. It would have saved us a lot of time.
I'm closing this discussion.

**Reply by Soothfast** [reply]
I started with whatever snap version was current in November 2023.  I wasn't aware until last week that xournalpp was updating automatically.  It does not notify me when it does.

As a novice I'm not sure how this could have played out differently.  You have been helpful though.
